README.md
# ReJSON Python Client

[![Build Status](https://travis-ci.org/RedisLabs/rejson-py.svg?branch=master)](https://travis-ci.org/RedisLabs/rejson-py)
[![Versions](https://img.shields.io/pypi/pyversions/rejson.svg)](https://github.com/RedisLabs/rejson-py)

rejson-py is a package that allows storing, updating and querying objects as
JSON documents in a [Redis](https://redis.io) database that is extended with the
[ReJSON module](https://github.com/redislabsmodules/rejson). The package extends
[redis-py](https://github.com/andymccurdy/redis-py)'s interface with ReJSON's
API, and performs on-the-fly serialization/deserialization of objects to/from
JSON.

## Installation

```bash
$ pip install rejson
```

## Usage example

```python
from rejson import Client, Path

rj = Client(host='localhost', port=6379)

# Set the key `obj` to some object
obj = {
    'answer': 42,
    'arr': [None, True, 3.14],
    'truth': {
        'coord': 'out there'
    }
}
rj.jsonset('obj', Path.rootPath(), obj)

# Get something
print 'Is there anybody... {}?'.format(
    rj.jsonget('obj', Path('.truth.coord'))
)

# Delete something (or perhaps nothing), append something and pop it
rj.jsondel('obj', Path('.arr[0]'))
rj.jsonarrappend('obj', Path('.arr'), 'something')
print '{} popped!'.format(rj.jsonarrpop('obj', Path('.arr')))

# Update something else
rj.jsonset('obj', Path('.answer'), 2.17)

# And use just like the regular redis-py client
jp = rj.pipeline()
jp.set('foo', 'bar')
jp.jsonset('baz', Path.rootPath(), 'qaz')
jp.execute()
```

## Encoding/Decoding

rejson-py uses Python's [`json`](https://docs.python.org/2/library/json.html).
The client can be set to use custom encoders/decoders at creation, or by calling
explicitly the [`setEncoder()`](API.md#setencoder) and
[`setDecoder()`](API.md#setdecoder) methods, respectively.

The following shows how to use this for a custom class that's stored as
a JSON string for example:

```python
from json import JSONEncoder, JSONDecoder
from rejson import Client

class CustomClass(object):
    "Some non-JSON-serializable"
    def __init__(self, s=None):
        if s is not None:
            # deserialize the instance from the serialization
            if s.startswith('CustomClass:'):
                ...
            else:
                raise Exception('unknown format')
        else:
            # initialize the instance
            ...

    def __str__(self):
        _str = 'CustomClass:'
        # append the instance's state to the serialization
        ...
        return _str

    ...

class CustomEncoder(JSONEncoder):
    "A custom encoder for the custom class"
    def default(self, obj):
        if isinstance(obj, CustomClass):
            return str(obj)
        return json.JSONEncoder.encode(self, obj)

class TestDecoder(JSONDecoder):
    "A custom decoder for the custom class"
    def decode(self, obj):
        d = json.JSONDecoder.decode(self, obj)
        if isinstance(d, basestring) and d.startswith('CustomClass:'):
            return CustomClass(d)
        return d

# Create a new instance of CustomClass
obj = CustomClass()

# Create a new client with the custom encoder and decoder
rj = Client(encoder=CustomEncoder(), decoder=CustomDecoder())

# Store the object
rj.jsonset('custom', Path.rootPath(), obj))

# Retrieve it
obj = rj.jsonget('custom', Path.rootPath())
```

## API

As rejson-py exposes the same methods as redis-py, it can be used as a drop-in
replacement. On top of Redis' core commands, the client also adds ReJSON's
vocabulary and a couple of helper methods. These are documented in the
[API.md](API.md) file, which can be generated by running:

```bash
$ python gendoc rejson > API.md
```

For complete documentation about ReJSON's commands, refer to
[rejson.io](http://rejson.io).

## License 

[BSD 2-Clause](LICENSE)


CHANGELOG.md
# rejson-py changelog

## 0.5.2 20191212

* #30 Use io.open() and 'utf-8' to read description

## 0.5.1 20191129

* #29 Use io.open() and 'utf-8' encoding to read version

## 0.5.0 20190901

* #26 Fixing retrieving non-ascii string with JSON.GET, added no_escape flag

## 0.4.0 20190308

* Add exception handler for .jsonget for non-existent keys

## 0.3.0 20190107

* Upgrade dependency to redis-py >=3.0.0

## 0.2.1 20170605

Recovery release

* Adds missing manifest

## 0.2 20170605

Codename "_[@hootnot](https://github.com/hootnot)_"

* Adds support for Python 3 and above, PEP8 and a lovely deployment framework via [#6](https://github.com/RedisLabs/rejson-py/pull/6) thanks to [@hootnot](https://github.com/hootnot)
* Adds coverage reporting (was %95 initially :))
* Turns coverage volume to 11

## 0.1 20170511

Initial release.


test/test.py
from __future__ import print_function
import six
import json
import unittest
from unittest import TestCase
from rejson import Client, Path


rj = None
port = 6379


class ReJSONTestCase(TestCase):

    def setUp(self):
        global rj
        rj = Client(port=port, decode_responses=True)
        rj.flushdb()

    def testJSONSetGetDelShouldSucceed(self):
        "Test basic JSONSet/Get/Del"

        self.assertTrue(rj.jsonset('foo', Path.rootPath(), 'bar'))
        self.assertEqual('bar', rj.jsonget('foo'))
        self.assertEqual(1, rj.jsondel('foo'))
        self.assertFalse(rj.exists('foo'))

    def testMGetShouldSucceed(self):
        "Test JSONMGet"

        rj.jsonset('1', Path.rootPath(), 1)
        rj.jsonset('2', Path.rootPath(), 2)
        r = rj.jsonmget(Path.rootPath(), '1', '2')
        e = [1, 2]
        self.assertListEqual(e, r)

    def testTypeShouldSucceed(self):
        "Test JSONType"

        rj.jsonset('1', Path.rootPath(), 1)
        self.assertEqual('integer', rj.jsontype('1'))

    def testNumIncrByShouldSucceed(self):
        "Test JSONNumIncrBy"

        rj.jsonset('num', Path.rootPath(), 1)
        self.assertEqual(2, rj.jsonnumincrby('num', Path.rootPath(), 1))
        self.assertEqual(2.5, rj.jsonnumincrby('num', Path.rootPath(), 0.5))
        self.assertEqual(1.25, rj.jsonnumincrby('num', Path.rootPath(), -1.25))

    def testNumMultByShouldSucceed(self):
        "Test JSONNumIncrBy"

        rj.jsonset('num', Path.rootPath(), 1)
        self.assertEqual(2, rj.jsonnummultby('num', Path.rootPath(), 2))
        self.assertEqual(5, rj.jsonnummultby('num', Path.rootPath(), 2.5))
        self.assertEqual(2.5, rj.jsonnummultby('num', Path.rootPath(), 0.5))

    def testStrAppendShouldSucceed(self):
        "Test JSONStrAppend"

        rj.jsonset('str', Path.rootPath(), 'foo')
        self.assertEqual(6, rj.jsonstrappend('str', 'bar', Path.rootPath()))
        self.assertEqual('foobar', rj.jsonget('str', Path.rootPath()))

    def testStrLenShouldSucceed(self):
        "Test JSONStrLen"

        rj.jsonset('str', Path.rootPath(), 'foo')
        self.assertEqual(3, rj.jsonstrlen('str', Path.rootPath()))
        rj.jsonstrappend('str', 'bar', Path.rootPath())
        self.assertEqual(6, rj.jsonstrlen('str', Path.rootPath()))

    def testArrAppendShouldSucceed(self):
        "Test JSONSArrAppend"

        rj.jsonset('arr', Path.rootPath(), [1])
        self.assertEqual(2, rj.jsonarrappend('arr', Path.rootPath(), 2))

    def testArrIndexShouldSucceed(self):
        "Test JSONSArrIndex"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(1, rj.jsonarrindex('arr', Path.rootPath(), 1))
        self.assertEqual(-1, rj.jsonarrindex('arr', Path.rootPath(), 1, 2))

    def testArrInsertShouldSucceed(self):
        "Test JSONSArrInsert"

        rj.jsonset('arr', Path.rootPath(), [0, 4])
        self.assertEqual(5, rj.jsonarrinsert('arr',
                                             Path.rootPath(), 1, *[1, 2, 3, ]))
        self.assertListEqual([0, 1, 2, 3, 4], rj.jsonget('arr'))

    def testArrLenShouldSucceed(self):
        "Test JSONSArrLen"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(5, rj.jsonarrlen('arr', Path.rootPath()))

    def testArrPopShouldSucceed(self):
        "Test JSONSArrPop"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(4, rj.jsonarrpop('arr', Path.rootPath(), 4))
        self.assertEqual(3, rj.jsonarrpop('arr', Path.rootPath(), -1))
        self.assertEqual(2, rj.jsonarrpop('arr', Path.rootPath()))
        self.assertEqual(0, rj.jsonarrpop('arr', Path.rootPath(), 0))
        self.assertListEqual([1], rj.jsonget('arr'))

    def testArrTrimShouldSucceed(self):
        "Test JSONSArrPop"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(3, rj.jsonarrtrim('arr', Path.rootPath(), 1, 3))
        self.assertListEqual([1, 2, 3], rj.jsonget('arr'))

    def testObjKeysShouldSucceed(self):
        "Test JSONSObjKeys"

        obj = {'foo': 'bar', 'baz': 'qaz'}
        rj.jsonset('obj', Path.rootPath(), obj)
        keys = rj.jsonobjkeys('obj', Path.rootPath())
        keys.sort()
        exp = [k for k in six.iterkeys(obj)]
        exp.sort()
        self.assertListEqual(exp, keys)

    def testObjLenShouldSucceed(self):
        "Test JSONSObjLen"

        obj = {'foo': 'bar', 'baz': 'qaz'}
        rj.jsonset('obj', Path.rootPath(), obj)
        self.assertEqual(len(obj), rj.jsonobjlen('obj', Path.rootPath()))

    def testPipelineShouldSucceed(self):
        "Test pipeline"

        p = rj.pipeline()
        p.jsonset('foo', Path.rootPath(), 'bar')
        p.jsonget('foo')
        p.jsondel('foo')
        p.exists('foo')
        self.assertListEqual([True, 'bar', 1, False], p.execute())

    def testCustomEncoderDecoderShouldSucceed(self):
        "Test a custom encoder and decoder"

        class CustomClass(object):
            key = ''
            val = ''

            def __init__(self, k='', v=''):
                self.key = k
                self.val = v

        class TestEncoder(json.JSONEncoder):
            def default(self, obj):
                if isinstance(obj, CustomClass):
                    return 'CustomClass:{}:{}'.format(obj.key, obj.val)
                return json.JSONEncoder.encode(self, obj)

        class TestDecoder(json.JSONDecoder):
            def decode(self, obj):
                d = json.JSONDecoder.decode(self, obj)
                if isinstance(d, six.string_types) and \
                        d.startswith('CustomClass:'):
                    s = d.split(':')
                    return CustomClass(k=s[1], v=s[2])
                return d

        rj = Client(encoder=TestEncoder(), decoder=TestDecoder(),
                    port=port, decode_responses=True)
        rj.flushdb()

        # Check a regular string
        self.assertTrue(rj.jsonset('foo', Path.rootPath(), 'bar'))
        self.assertEqual('string', rj.jsontype('foo', Path.rootPath()))
        self.assertEqual('bar', rj.jsonget('foo', Path.rootPath()))

        # Check the custom encoder
        self.assertTrue(rj.jsonset('cus', Path.rootPath(),
                                   CustomClass('foo', 'bar')))
        obj = rj.jsonget('cus', Path.rootPath())
        self.assertIsNotNone(obj)
        self.assertEqual(CustomClass, obj.__class__)
        self.assertEqual('foo', obj.key)
        self.assertEqual('bar', obj.val)

    def testUsageExampleShouldSucceed(self):
        "Test the usage example"

        # Create a new rejson-py client
        rj = Client(host='localhost', port=port, decode_responses=True)

        # Set the key `obj` to some object
        obj = {
            'answer': 42,
            'arr': [None, True, 3.14],
            'truth': {
                'coord': 'out there'
            }
        }
        rj.jsonset('obj', Path.rootPath(), obj)

        # Get something
        print('Is there anybody... {}?'.format(
            rj.jsonget('obj', Path('.truth.coord'))
        ))

        # Delete something (or perhaps nothing), append something and pop it
        rj.jsondel('obj', Path('.arr[0]'))
        rj.jsonarrappend('obj', Path('.arr'), 'something')
        print('{} popped!'.format(rj.jsonarrpop('obj', Path('.arr'))))

        # Update something else
        rj.jsonset('obj', Path('.answer'), 2.17)

        # And use just like the regular redis-py client
        jp = rj.pipeline()
        jp.set('foo', 'bar')
        jp.jsonset('baz', Path.rootPath(), 'qaz')
        jp.execute()


if __name__ == '__main__':
    unittest.main()


tests/test_rejson.py
import six
import json
import unittest
from unittest import TestCase
from rejson import Client, Path

rj = None
port = 6379


class ReJSONTestCase(TestCase):

    def setUp(self):
        global rj
        rj = Client(port=port, decode_responses=True)
        rj.flushdb()

    def testJSONSetGetDelShouldSucceed(self):
        "Test basic JSONSet/Get/Del"

        self.assertTrue(rj.jsonset('foo', Path.rootPath(), 'bar'))
        self.assertEqual('bar', rj.jsonget('foo'))
        self.assertEqual(None, rj.jsonget('baz'))
        self.assertEqual(1, rj.jsondel('foo'))
        self.assertFalse(rj.exists('foo'))

    def testJSONSetGetDelNonAsciiShouldSucceed(self):
        "Test non-ascii JSONSet/Get/Del"

        self.assertTrue(rj.jsonset('notascii', Path.rootPath(), 'hyvää-élève'))
        self.assertNotEqual('hyvää-élève', rj.jsonget('notascii'))
        self.assertEqual('hyvää-élève', rj.jsonget('notascii', no_escape=True))
        self.assertEqual(1, rj.jsondel('notascii'))
        self.assertFalse(rj.exists('notascii'))

    def testJSONSetExistentialModifiersShouldSucceed(self):
        "Test JSONSet's NX/XX flags"

        obj = { 'foo': 'bar' }
        self.assertTrue(rj.jsonset('obj', Path.rootPath(), obj))

        # Test that flags prevent updates when conditions are unmet
        self.assertFalse(rj.jsonset('obj', Path('foo'), 'baz', nx=True))
        self.assertFalse(rj.jsonset('obj', Path('qaz'), 'baz', xx=True))

        # Test that flags allow updates when conditions are met
        self.assertTrue(rj.jsonset('obj', Path('foo'), 'baz', xx=True))
        self.assertTrue(rj.jsonset('obj', Path('qaz'), 'baz', nx=True))

        # Test that flags are mutually exlusive
        with self.assertRaises(Exception) as context:
            rj.jsonset('obj', Path('foo'), 'baz', nx=True, xx=True)

    def testMGetShouldSucceed(self):
        "Test JSONMGet"

        rj.jsonset('1', Path.rootPath(), 1)
        rj.jsonset('2', Path.rootPath(), 2)
        r = rj.jsonmget(Path.rootPath(), '1', '2')
        e = [1, 2]
        self.assertListEqual(e, r)

    def testTypeShouldSucceed(self):
        "Test JSONType"

        rj.jsonset('1', Path.rootPath(), 1)
        self.assertEqual('integer', rj.jsontype('1'))

    def testNumIncrByShouldSucceed(self):
        "Test JSONNumIncrBy"

        rj.jsonset('num', Path.rootPath(), 1)
        self.assertEqual(2, rj.jsonnumincrby('num', Path.rootPath(), 1))
        self.assertEqual(2.5, rj.jsonnumincrby('num', Path.rootPath(), 0.5))
        self.assertEqual(1.25, rj.jsonnumincrby('num', Path.rootPath(), -1.25))

    def testNumMultByShouldSucceed(self):
        "Test JSONNumIncrBy"

        rj.jsonset('num', Path.rootPath(), 1)
        self.assertEqual(2, rj.jsonnummultby('num', Path.rootPath(), 2))
        self.assertEqual(5, rj.jsonnummultby('num', Path.rootPath(), 2.5))
        self.assertEqual(2.5, rj.jsonnummultby('num', Path.rootPath(), 0.5))

    def testStrAppendShouldSucceed(self):
        "Test JSONStrAppend"

        rj.jsonset('str', Path.rootPath(), 'foo')
        self.assertEqual(6, rj.jsonstrappend('str', 'bar', Path.rootPath()))
        self.assertEqual('foobar', rj.jsonget('str', Path.rootPath()))

    def testStrLenShouldSucceed(self):
        "Test JSONStrLen"

        rj.jsonset('str', Path.rootPath(), 'foo')
        self.assertEqual(3, rj.jsonstrlen('str', Path.rootPath()))
        rj.jsonstrappend('str', 'bar', Path.rootPath())
        self.assertEqual(6, rj.jsonstrlen('str', Path.rootPath()))

    def testArrAppendShouldSucceed(self):
        "Test JSONSArrAppend"

        rj.jsonset('arr', Path.rootPath(), [1])
        self.assertEqual(2, rj.jsonarrappend('arr', Path.rootPath(), 2))

    def testArrIndexShouldSucceed(self):
        "Test JSONSArrIndex"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(1, rj.jsonarrindex('arr', Path.rootPath(), 1))
        self.assertEqual(-1, rj.jsonarrindex('arr', Path.rootPath(), 1, 2))

    def testArrInsertShouldSucceed(self):
        "Test JSONSArrInsert"

        rj.jsonset('arr', Path.rootPath(), [0, 4])
        self.assertEqual(5, rj.jsonarrinsert('arr',
                                             Path.rootPath(), 1, *[1, 2, 3, ]))
        self.assertListEqual([0, 1, 2, 3, 4], rj.jsonget('arr'))

    def testArrLenShouldSucceed(self):
        "Test JSONSArrLen"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(5, rj.jsonarrlen('arr', Path.rootPath()))

    def testArrPopShouldSucceed(self):
        "Test JSONSArrPop"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(4, rj.jsonarrpop('arr', Path.rootPath(), 4))
        self.assertEqual(3, rj.jsonarrpop('arr', Path.rootPath(), -1))
        self.assertEqual(2, rj.jsonarrpop('arr', Path.rootPath()))
        self.assertEqual(0, rj.jsonarrpop('arr', Path.rootPath(), 0))
        self.assertListEqual([1], rj.jsonget('arr'))

    def testArrTrimShouldSucceed(self):
        "Test JSONSArrPop"

        rj.jsonset('arr', Path.rootPath(), [0, 1, 2, 3, 4])
        self.assertEqual(3, rj.jsonarrtrim('arr', Path.rootPath(), 1, 3))
        self.assertListEqual([1, 2, 3], rj.jsonget('arr'))

    def testObjKeysShouldSucceed(self):
        "Test JSONSObjKeys"

        obj = {'foo': 'bar', 'baz': 'qaz'}
        rj.jsonset('obj', Path.rootPath(), obj)
        keys = rj.jsonobjkeys('obj', Path.rootPath())
        keys.sort()
        exp = [k for k in six.iterkeys(obj)]
        exp.sort()
        self.assertListEqual(exp, keys)

    def testObjLenShouldSucceed(self):
        "Test JSONSObjLen"

        obj = {'foo': 'bar', 'baz': 'qaz'}
        rj.jsonset('obj', Path.rootPath(), obj)
        self.assertEqual(len(obj), rj.jsonobjlen('obj', Path.rootPath()))

    def testPipelineShouldSucceed(self):
        "Test pipeline"

        p = rj.pipeline()
        p.jsonset('foo', Path.rootPath(), 'bar')
        p.jsonget('foo')
        p.jsondel('foo')
        p.exists('foo')
        self.assertListEqual([True, 'bar', 1, False], p.execute())

    def testCustomEncoderDecoderShouldSucceed(self):
        "Test a custom encoder and decoder"

        class CustomClass(object):
            key = ''
            val = ''

            def __init__(self, k='', v=''):
                self.key = k
                self.val = v

        class TestEncoder(json.JSONEncoder):
            def default(self, obj):
                if isinstance(obj, CustomClass):
                    return 'CustomClass:{}:{}'.format(obj.key, obj.val)
                return json.JSONEncoder.encode(self, obj)

        class TestDecoder(json.JSONDecoder):
            def decode(self, obj):
                d = json.JSONDecoder.decode(self, obj)
                if isinstance(d, six.string_types) and \
                        d.startswith('CustomClass:'):
                    s = d.split(':')
                    return CustomClass(k=s[1], v=s[2])
                return d

        rj = Client(encoder=TestEncoder(), decoder=TestDecoder(),
                    port=port, decode_responses=True)
        rj.flushdb()

        # Check a regular string
        self.assertTrue(rj.jsonset('foo', Path.rootPath(), 'bar'))
        self.assertEqual('string', rj.jsontype('foo', Path.rootPath()))
        self.assertEqual('bar', rj.jsonget('foo', Path.rootPath()))

        # Check the custom encoder
        self.assertTrue(rj.jsonset('cus', Path.rootPath(),
                                   CustomClass('foo', 'bar')))
        obj = rj.jsonget('cus', Path.rootPath())
        self.assertIsNotNone(obj)
        self.assertEqual(CustomClass, obj.__class__)
        self.assertEqual('foo', obj.key)
        self.assertEqual('bar', obj.val)

    def testUsageExampleShouldSucceed(self):
        "Test the usage example"

        # Create a new rejson-py client
        rj = Client(host='localhost', port=port, decode_responses=True)

        # Set the key `obj` to some object
        obj = {
            'answer': 42,
            'arr': [None, True, 3.14],
            'truth': {
                'coord': 'out there'
            }
        }
        rj.jsonset('obj', Path.rootPath(), obj)

        # Get something
        rv = rj.jsonget('obj', Path('.truth.coord'))
        self.assertEqual(obj['truth']['coord'], rv)

        # Delete something (or perhaps nothing), append something and pop it
        value = "something"
        rj.jsondel('obj', Path('.arr[0]'))
        rj.jsonarrappend('obj', Path('.arr'), value)
        rv = rj.jsonarrpop('obj', Path('.arr'))
        self.assertEqual(value, rv)

        # Update something else
        value = 2.17
        rj.jsonset('obj', Path('.answer'), value)
        rv = rj.jsonget('obj', Path('.answer'))
        self.assertEqual(value, rv)

        # And use just like the regular redis-py client
        jp = rj.pipeline()
        jp.set('foo', 'bar')
        jp.jsonset('baz', Path.rootPath(), 'qaz')
        jp.execute()
        rv1 = rj.get('foo')
        self.assertEqual('bar', rv1)
        rv2 = rj.jsonget('baz')
        self.assertEqual('qaz', rv2)


if __name__ == '__main__':
    unittest.main()


gendoc.py
import pydoc
import os, sys

module_header = "# Package {} Documentation\n"
class_header = "## Class {}"
function_header = "### {}"

def getmarkdown(module):
    output = [ module_header.format(module.__name__) ]
    if module.__doc__:
        output.append(module.__doc__)
    output.extend(getclasses(module))
    return "\n".join(str(x) for x in output)

def getclasses(item):
    output = list()
    for cl in pydoc.inspect.getmembers(item, pydoc.inspect.isclass):
        if cl[0] != "__class__" and not cl[0].startswith("_"):
            # Consider anything that starts with _ private
            # and don't document it
            output.append(class_header.format(cl[0])) 
            # Get the docstring
            output.append(pydoc.inspect.getdoc(cl[1]))
            # Get the functions
            output.extend(getfunctions(cl[1]))
            # Recurse into any subclasses
            output.extend(getclasses(cl[1]))
            output.append('\n')
    return output

def getfunctions(item):
    output = list()
    bcfuncs = set()
    # Get the base class functions
    for bc in item.__bases__:
        for func in pydoc.inspect.getmembers(bc, pydoc.inspect.ismethod):
            if func[0].startswith('_'):
                continue
            bcfuncs.add(func[0])
    #print item
    for func in pydoc.inspect.getmembers(item, pydoc.inspect.ismethod):
        if (func[0].startswith('_') and func[0] != '__init__') or func[0] in bcfuncs:
            continue
        output.append(function_header.format(func[0].replace('_', '\\_')))
        # Get the signature
        output.append ('```py\n')
        output.append('def %s%s\n' % (func[0], pydoc.inspect.formatargspec(*pydoc.inspect.getargspec(func[1]))))
        output.append ('```\n')
        # get the docstring
        if pydoc.inspect.getdoc(func[1]):
            output.append('\n')
            output.append(pydoc.inspect.getdoc(func[1]))
        output.append('\n')
    return output

def generatedocs(module):
    try:
        sys.path.append(os.getcwd())
        # Attempt import
        mod = pydoc.safeimport(module)
        if mod is None:
           print("Module not found")
        # Module imported correctly, let's create the docs
        return getmarkdown(mod)
    except pydoc.ErrorDuringImport as e:
        print("Error while trying to import " + module)

if __name__ == '__main__':
    # print generatedocs(sys.argv[1])
    print generatedocs('rejson')


rejson/__init__.py
"""
rejson-py is a package that allows storing, updating and querying objects as
JSON documents in a [Redis](https://redis.io) database that is extended with the
[ReJSON module](https://github.com/redislabsmodules/rejson). The package extends
[redis-py](https://github.com/andymccurdy/redis-py)'s interface with ReJSON's
API, and performs on-the-fly serialization/deserialization of objects to/from
JSON.

## Installation

```bash
$ pip install rejson
```

## Usage example

```python
from rejson import Client, Path

rj = Client(host='localhost', port=6379)

# Set the key `obj` to some object
obj = {
    'answer': 42,
    'arr': [None, True, 3.14],
    'truth': {
        'coord': 'out there'
    }
}
rj.jsonset('obj', Path.rootPath(), obj)

# Get something
print 'Is there anybody... {}?'.format(
    rj.jsonget('obj', Path('.truth.coord'))
)

# Delete something (or perhaps nothing), append something and pop it
rj.jsondel('obj', Path('.arr[0]'))
rj.jsonarrappend('obj', Path('.arr'), 'something')
print '{} popped!'.format(rj.jsonarrpop('obj', Path('.arr')))

# Update something else
rj.jsonset('obj', Path('.answer'), 2.17)

# And use just like the regular redis-py client
jp = rj.pipeline()
jp.set('foo', 'bar')
jp.jsonset('baz', Path.rootPath(), 'qaz')
jp.execute()

# If you use non-ascii character in your JSON data, you can add the no_escape flag to JSON.GET command
obj_non_ascii = {
    'non_ascii_string': 'hyvää'
    }

rj.jsonset('non-ascii', Path.rootPath(), obj_non_ascii)
print '{} is a non-ascii string'.format(rj.jsonget('non-ascii', Path('.non_ascii_string'), no_escape=True))
```

## Encoding/Decoding

rejson-py uses Python's [`json`](https://docs.python.org/2/library/json.html).
The client can be set to use custom encoders/decoders at creation, or by calling
explicitly the [`setEncoder()`](API.md#setencoder) and
[`setDecoder()`](API.md#setdecoder) methods, respectively.

The following shows how to use this for a custom class that's stored as
a JSON string for example:

```python
from json import JSONEncoder, JSONDecoder
from rejson import Client

class CustomClass(object):
    "Some non-JSON-serializable"
    def __init__(self, s=None):
        if s is not None:
            # deserialize the instance from the serialization
            if s.startswith('CustomClass:'):
                ...
            else:
                raise Exception('unknown format')
        else:
            # initialize the instance
            ...

    def __str__(self):
        _str = 'CustomClass:'
        # append the instance's state to the serialization
        ...
        return _str

    ...

class CustomEncoder(JSONEncoder):
    "A custom encoder for the custom class"
    def default(self, obj):
        if isinstance(obj, CustomClass):
            return str(obj)
        return json.JSONEncoder.encode(self, obj)

class TestDecoder(JSONDecoder):
    "A custom decoder for the custom class"
    def decode(self, obj):
        d = json.JSONDecoder.decode(self, obj)
        if isinstance(d, basestring) and d.startswith('CustomClass:'):
            return CustomClass(d)
        return d

# Create a new instance of CustomClass
obj = CustomClass()

# Create a new client with the custom encoder and decoder
rj = Client(encoder=CustomEncoder(), decoder=CustomDecoder())

# Store the object
rj.jsonset('custom', Path.rootPath(), obj))

# Retrieve it
obj = rj.jsonget('custom', Path.rootPath())
```
"""
__version__ = "0.5.2"
from .client import Client
from .path import Path
